qtwidgetsextra/filesystem/QFileSystemListView.cpp

```diff
@@ -0,0 +1,257 @@
+        QMap<QModelIndex, QMap<QModelIndex, int> > mapIndexes;
+        QItemSelection itemSelection;
+
+        foreach (const QString &item, selection) {
+            const QModelIndex index = model->index(item, column);
+
+            if (index.isValid()) {
+                mapIndexes[index.parent()][index]++;
+            }
+        }
+
+        foreach (const QModelIndex &parent, mapIndexes.keys()) {
+            const QModelIndexList indexes = mapIndexes[parent].keys();
+            int top = -1;
+            int bottom = -1;
+
+            for (int i = 0; i < indexes.count(); i++) {
+                const QModelIndex &index = indexes[i];
+                const bool hasNext = i +1 < indexes.count();
+
+                if (top ==-1) {
+                    top = index.row();
+                }
+                else {
+                    if (bottom == -1) {
+                        // index is not contiguous
+                        if (index.row() != top +1) {
+                            const QModelIndex idx = model->index(top, column, parent);
+                            itemSelection << QItemSelectionRange(idx, idx);
+
+                            top = index.row();
+                            bottom = -1;
+                        }
+                        else {
+                            bottom = index.row();
+                        }
+                    }
+                    else if (index.row() != bottom +1) {
+                        // index is not contiguous
+                        if (index.row() != top +1) {
+                            const QModelIndex idx1 = model->index(top, column, parent);
+                            const QModelIndex idx2 = model->index(bottom, column, parent);
+                            itemSelection << QItemSelectionRange(idx1, idx2);
+
+                            top = index.row();
+                            bottom = -1;
+                        }
+                        else {
+                            bottom = index.row();
+                        }
+                    }
+                }
+
+                if (!hasNext && top != -1) {
+                    if (bottom == -1) {
+                        bottom = top;
+                    }
+
+                    const QModelIndex idx1 = model->index(top, column, parent);
+                    const QModelIndex idx2 = model->index(bottom, column, parent);
+                    itemSelection << QItemSelection(idx1, idx2);
+
+                    top = -1;
+                    bottom = -1;
+                }
+            }
+        }
+
+        widget->selectionModel()->select(itemSelection, flags);
+    }
+
+public slots:
+    void activated(const QModelIndex &index) {
+        if (!widget->browsable() || !model->isDir(index) || QApplication::keyboardModifiers() != Qt::NoModifier) {
+            return;
+        }
+
+        const QString filePath = model->filePath(index);
+        widget->setRootPath(filePath);
+    }
+
+public:
```

Comment:
Yeah I don't like much putting in the same public scope members and functions.
I will check Qt convention.